src/Aspire.Hosting/ResourceBuilderExtensions.cs

```diff
@@ -1139,6 +1140,289 @@ public static IResourceBuilder<T> WithCommand<T>(
+        {
+            var rns = e.Services.GetRequiredService<ResourceNotificationService>();
+            var watchTask = Task.Run(async () =>
+            {
+                await foreach (var resourceEvent in rns.WatchAsync(ct).WithCancellation(ct))
+                {
+                    if (resourceEvent.Resource == endpoint.Resource)
+                    {
+                        var resourceState = resourceEvent.Snapshot.State?.Text;
+                        targetRunning = resourceState == KnownResourceStates.Running || resourceState == KnownResourceStates.RuntimeUnhealthy;
+                    }
+                }
+            }, ct);
+
+            return Task.CompletedTask;
+        });
+
+        builder.WithCommand(commandName, displayName,
+            async context =>
+            {
+                if (!endpoint.IsAllocated)
+                {
+                    return new ExecuteCommandResult { Success = false, ErrorMessage = "Endpoints are not yet allocated." };
+                }
+
+                var uri = new UriBuilder(endpoint.Url) { Path = path }.Uri;
+                var httpClient = context.ServiceProvider.GetRequiredService<IHttpClientFactory>().CreateClient(httpClientName ?? Options.DefaultName);
+                var request = new HttpRequestMessage(method, uri);
+                if (configureRequest is not null)
+                {
+                    await configureRequest(request).ConfigureAwait(false);
+                }
+                try
+                {
+                    var response = await httpClient.SendAsync(request, context.CancellationToken).ConfigureAwait(false);
+                    if (getCommandResult is not null)
+                    {
+                        return await getCommandResult(response).ConfigureAwait(false);
+                    }
+
+                    response.EnsureSuccessStatusCode();
+                    return CommandResults.Success();
+                }
+                catch (Exception ex)
+                {
+                    return CommandResults.Failure(ex);
+                }
+            },
+            updateState: context => targetRunning ? ResourceCommandState.Enabled : ResourceCommandState.Disabled,
+            displayDescription: displayDescription,
+            confirmationMessage: confirmationMessage,
+            iconName: iconName,
+            iconVariant: iconVariant,
+            isHighlighted: isHighlighted);
+
+        return builder;
+    }
+
+    // These match the default endpoint names resulting from calling WithHttpsEndpoint or WithHttpEndpoint as well as the defaults
+    // created for ASP.NET Core projects with the default launch settings added via AddProject. HTTPS is first so that we prefer it
+    // if found.
+    private static readonly string[] s_httpSchemes = ["https", "http"];
+
+    private static Func<EndpointReference> PreferredEndpointSelector<TResource>(IResourceBuilder<TResource> builder, string[]? endpointNames = null)
```

Comment:
(super nit) - I wonder if this method makes sense all together. It has 2 modes: either you supplied a list of endpointNames, or you didn't. It doesn't really have common code between those 2 cases. Could split it into 2 methods, one that takes a `string[]` and one that doesn't.

Reply:
Yeah it used to fall through if you supplied names to looking for the default named ones, but that's not the case anymore.

Reply:
OK I split it up into `NamedEndpointSelector` and `DefaultEndpointSelector`